### PR TITLE
Ingest correctly handle upsert operations and drop processors together

### LINKED ISSUE #36746: ingest: upsert with drop processor returns error
If performing an upsert with a default pipeline, and that default pipeline drops the document, an error will be returned to the client. 

This is a minor bug since the end result is the same, the document is not indexed. However, it can pollute error logs, cause confusion, and have a minor performance penalty due to the exception and logging. 

Both normal upsert and bulk upsert will error, but with different signatures
```
DELETE test
PUT test
{
  "settings": {
    "index.default_pipeline": "dropper"
  }
}
PUT _ingest/pipeline/dropper
{
  "processors": [
    {
      "drop": {}
    }
  ]
}
```
Non-bulk upsert:
```
POST test/doc/1/_update
{
  "script":{
    "source": "ctx._source.foo = 'bar'" 
  },
  "upsert" :{
    "foo" : "bar"
  }
}
```
Results in 
```
{
  "error": {
    "root_cause": [
      {
        "type": "remote_transport_exception",
        "reason": "[node-0][127.0.0.1:9300][indices:data/write/update[s]]"
      }
    ],
    "type": "class_cast_exception",
    "reason": "class org.elasticsearch.action.update.UpdateResponse cannot be cast to class org.elasticsearch.action.index.IndexResponse (org.elasticsearch.action.update.UpdateResponse and org.elasticsearch.action.index.IndexResponse are in unnamed module of loader 'app')"
  },
  "status": 500
}
```
and bulk upsert results in
```
    {
      "index" : {
        "_index" : null,
        "_type" : null,
        "_id" : null,
        "status" : 500,
        "error" : {
          "type" : "null_pointer_exception",
          "reason" : null
        }
      }
    }
```

### PR DESCRIPTION
**Pull request by joegallo**
Closes #36746

`new ShardId(String index, ...)` checks that the `String` arguments it receives are not `null`. When an upsert and a `drop` processor meet, the thing returned from `getIndexWriteRequest(...)` won't have the index populated, so we end up passing `null` where it can't go.

But we don't need to call `getIndexWriteRequest(...)` at all! The `docWriteRequest` argument to it already has everything we need **or so I think** (feel free to dis-abuse me of this notion).

~WIP because I have cleared all out my test changes and added some extra asserts. Once I get a green run from Jenkins on that, I'll revert the revert and get my new tests back (but lose the extra asserts). After that I might still add a test or two.~

### CODE REVIEW

**Comment by joegallo** [comment]
Screenshot 2024-01-20 at 12 44 19 PM

15/15, so I'm adding the tests back now.

**Comment by masseyke** [comment]
Seems like a good idea. One test that would be good to have (and this might be one you already have in mind) is a non-bulk upsert since the issue reports a different stack trace for non-bulk. I think the current tests only cover bulk, unless I missed something.

**Comment by joegallo** [comment]
>  One test that would be good to have [...]

Yes! That's the precisely the test I have in mind.

**Comment by joegallo** [comment]
@elasticmachine run elasticsearch-ci/packaging-tests-windows-sample

**Comment by joegallo** [comment]
A bunch of the lines of code changed here were just whitespace twiddling in the tests, I'd recommend reviewing this PR with [`?w=1`](https://github.com/elastic/elasticsearch/pull/104585/files?w=1) to ignore all that.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 1]
```diff
@@ -1110,35 +1109,31 @@ private void assertResponsesAreCorrect(BulkItemResponse[] bulkResponses, BulkIte
...
-            itemResponses.add(BulkItemResponse.failure(slot, indexRequest.opType(), failure));
+            BulkItemResponse.Failure failure = new BulkItemResponse.Failure(docWriteRequest.index(), id, e);
+            itemResponses.add(BulkItemResponse.failure(slot, docWriteRequest.opType(), failure));
+        }
+
+        synchronized void markItemAsDropped(int slot) {
+            final DocWriteRequest<?> docWriteRequest = bulkRequest.requests().get(slot);
+            final String id = Objects.requireNonNullElse(docWriteRequest.id(), DROPPED_OR_FAILED_ITEM_WITH_AUTO_GENERATED_ID);
+            failedSlots.set(slot);
+            UpdateResponse dropped = new UpdateResponse(
```
Comment (masseyke) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:1128]: Why is this an UpdateResponse? I see that it was previously, but it leads to some weird things, like getting an UpdateResponse from an IndexRequest (which you also had to account for in this PR).

Comment (joegallo) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:1128]: I don't think IndexResponse lets us express anything except `CREATED` or `UPDATED`:

```
    private static Result assertCreatedOrUpdated(Result result) {
        assert result == Result.CREATED || result == Result.UPDATED;
        return result;
    }
```

But yeah, it is inconvenient (and also ancient!).

**Review by masseyke — APPROVED** [review]
LGTM

**Pull request merged**